File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        
        if len(nums2) < len(nums1):
            nums1, nums2 = nums2, nums1
        
        total = len(nums1) + len(nums2)
        half = total // 2
        
        l, r = 0, len(nums1) - 1
        while True:
            i = (l + r) // 2    # for nums1
            j = half - i - 2    # for nums2
            
            left_nums1 = nums1[i] if i >= 0 else float("-inf")
            right_nums1 = nums1[i+1] if (i + 1) < len(nums1) else float("inf")
            
            left_nums2 = nums2[j] if j >= 0 else float("-inf")
            right_nums2 = nums2[j+1] if (j + 1) < len(nums2) else float("inf")
            
            if left_nums1 <= right_nums2 and left_nums2 <= right_nums1:
                
                if total % 2:
                    return min(right_nums1, right_nums2)
                else:
                    return (max(left_nums1, left_nums2) + min(right_nums1, right_nums2)) / 2
            
            elif left_nums1 > right_nums2:
                r = i - 1
            else:
                l = i + 1

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int n1 = nums1.size(), n2 = nums2.size();
        if(n1 > n2){
            return findMedianSortedArrays(nums2,nums1);
        }
        int l = 0, r = n1;
        int onLeft = (n1+n2+1)/2;
        int n = n1 + n2;
        while(l<=r){
            int mid1 = (l+r)/2, mid2 = onLeft - mid1;
            int l1 = INT_MIN, l2 = INT_MIN;
            int r1 = INT_MAX, r2 = INT_MAX;
            if(mid1<n1){
                r1 = nums1[mid1];
            }
            if(mid2<n2){
                r2 = nums2[mid2];
            }
            if(mid1 - 1 >= 0){
                l1 = nums1[mid1-1];
            }
            if(mid2 - 1 >= 0){
                l2 = nums2[mid2-1];
            }
            if(l1<=r2 && l2<=r1){
                if(n%2 == 1){
                    return max(l1,l2);
                }
                else{
                    return ((double)(max(l1,l2)+min(r1,r2)))/2.0;
                }
            }
            if(l1>r2){
                r = mid1 - 1;
            }
            else{
                l = mid1 + 1;
            }
        }
        return 0.0;
    }
};